File: src/main/java/workshop/part4/VerdictCache.java
package workshop.part4;

import java.util.concurrent.ConcurrentHashMap;

import javaslang.collection.Stream;
import javaslang.control.Option;
import workshop.common.ad.Ad;
import workshop.part1.Verdict;

public class VerdictCache {

    private ConcurrentHashMap<Ad, Verdict> cache = new ConcurrentHashMap<>();

    public Option<Verdict> get(Ad ad) {
        return Option.of(cache.get(ad));
    }

    public Option<Verdict> get(Integer adId) {
        return Stream.ofAll(cache.keySet())
                .find(a -> a.getAdId().equals(adId))
                .map(cache::get);
    }

    public void put(Ad ad, Verdict verdict) {
        cache.put(ad, verdict);
    }
}

File: src/main/kotlin/workshop/part4/VerdictCache.kt
package workshop.part4

import javaslang.collection.Stream
import javaslang.control.Option
import workshop.common.ad.Ad
import workshop.part1.Verdict
import java.util.concurrent.ConcurrentHashMap

open class VerdictCache {

    private val cache = ConcurrentHashMap<Ad, Verdict>()

    operator fun get(ad: Ad): Option<Verdict> {
        return Option.of(cache[ad])
    }

    open operator fun get(adId: Int?): Option<Verdict> {
        return Stream.ofAll(cache.keys)
                .find { a -> a.adId == adId }
                .map({ cache[it] })
    }

    fun put(ad: Ad, verdict: Verdict) {
        cache.put(ad, verdict)
    }
}
